src/Storages/Statistics/ConditionSelectivityEstimator.cpp

```diff
@@ -9,112 +15,142 @@ namespace ErrorCodes
+bool ConditionSelectivityEstimator::extractAtomFromTree(const RPNBuilderTreeNode & node, RPNElement & out) const
 {
-    auto result = tryToExtractSingleColumn(node);
-    if (result.second != 1)
-        return default_unknown_cond_factor * total_rows;
-
-    String column = result.first;
-    auto it = column_estimators.find(column);
+    const auto * node_dag = node.getDAGNode();
+    if (node_dag && node_dag->result_type->equals(DataTypeNullable(std::make_shared<DataTypeNothing>())))
+    {
+        /// If the inferred result type is Nullable(Nothing) at the query analysis stage,
+        /// we don't analyze this node further as its condition will always be false.
+        out.function = RPNElement::ALWAYS_FALSE;
+        return true;
+    }
+    Field const_value;
+    DataTypePtr const_type;
+    String column_name;
 
-    /// If there the estimator of the column is not found or there are no data at all,
-    /// we use dummy estimation.
-    bool dummy = false;
-    ColumnSelectivityEstimator estimator;
-    if (it != column_estimators.end())
-        estimator = it->second;
-    else
-        dummy = true;
+    if (node.isFunction())
+    {
+        auto func = node.toFunctionNode();
+        size_t num_args = func.getArgumentsSize();
 
-    auto [op, val] = extractBinaryOp(node, column);
+        String func_name = func.getFunctionName();
+        if (atom_map.find(func_name) == std::end(atom_map))
+            return false;
 
-    if (dummy)
-    {
-        if (op == "equals")
-            return default_cond_equal_factor * total_rows;
-        if (op == "less" || op == "lessOrEquals" || op == "greater" || op == "greaterOrEquals")
-            return default_cond_range_factor * total_rows;
-        return default_unknown_cond_factor * total_rows;
+        if (num_args == 2)
+        {
+            if (func.getArgumentAt(1).tryGetConstant(const_value, const_type))
+            {
+                if (const_value.isNull())
+                {
+                    out.function = RPNElement::ALWAYS_FALSE;
+                    return true;
+                }
+                column_name = func.getArgumentAt(0).getColumnName();
+            }
+            else if (func.getArgumentAt(0).tryGetConstant(const_value, const_type))
+            {
+                if (const_value.isNull())
+                {
+                    out.function = RPNElement::ALWAYS_FALSE;
+                    return true;
+                }
+                column_name = func.getArgumentAt(1).getColumnName();
+                if (func_name == "less")
+                    func_name = "greater";
+                else if (func_name == "greater")
+                    func_name = "less";
+                else if (func_name == "greaterOrEquals")
+                    func_name = "lessOrEquals";
+                else if (func_name == "lessOrEquals")
+                    func_name = "greaterOrEquals";
```

Comment:
Why is the function renamed? Did we swap arguments?

Reply:
we analyze `5 < a` as `a > 5` to generate a correct range `(5, inf)`, otherwise it will be `(-inf, 5)`. I copied this code from https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/KeyCondition.cpp#L2203